### PR TITLE
format fetchedges

### PR DESCRIPTION
**Pull request by nevermore3**
1、refactor fetch edges
2、support FETCH PROP ON edgeName "A"->"B" YIELD edge AS e, src(edge), dst(edge), rank(edge), type(edge)

### CODE REVIEW

**Comment by nevermore3** [comment]
subjob of https://github.com/vesoft-inc/nebula/issues/2594

File: tests/tck/features/delete/DeleteTag.IntVid.feature [hunk 1]
```diff
@@ -2,6 +2,7 @@
 #
 # This source code is licensed under Apache 2.0 License,
 # attached with Common Clause Condition 1.0, found in the LICENSES directory.
+@jmq
```
Comment (Shylock-Hg) [inline tests/tck/features/delete/DeleteTag.IntVid.feature:5]: Why add this?

File: tests/tck/features/fetch/FetchEdges.intVid.feature [hunk 2]
```diff
@@ -22,17 +22,17 @@ Feature: Fetch Int Vid Edges
     # Fetch prop on an edge without yield
     When executing query:
       """
-      FETCH PROP ON serve hash("Boris Diaw")->hash("Spurs")
+      FETCH PROP ON serve hash("Boris Diaw")->hash("Spurs") YIELD edge
```
Comment (Shylock-Hg) [inline tests/tck/features/fetch/FetchEdges.intVid.feature:25]: Don't modify the origin case if it's compatible.

File: tests/tck/features/delete/DeleteTag.IntVid.feature (hunk 1 above)
Comment (nevermore3) [reply tests/tck/features/delete/DeleteTag.IntVid.feature:5]: not done,just test

File: tests/tck/features/fetch/FetchEdges.intVid.feature (hunk 2 above)
Comment (nevermore3) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:25]: will fix it

File: tests/tck/features/fetch/FetchEdges.strVid.feature [hunk 3]
```diff
@@ -249,3 +256,64 @@ Feature: Fetch String Vid Edges
...
+      | "Boris Diaw" | "Spurs"    | 0           | [:serve "Boris Diaw"->"Spurs" @0 {end_year: 2016, start_year: 2012}] | "Boris Diaw" | "Spurs"  | "serve" | 0    |
+    When executing query:
+      """
+      FETCH PROP ON serve "Boris Diaw"->"Not Exist" YIELD src(edge) as a
+      """
+    Then the result should be, in any order:
+      | serve._src | serve._dst | serve._rank | a |
+    When executing query:
+      """
+      FETCH PROP ON like "Tony Parker"->"Tim Duncan", "Grant Hill" -> "Tracy McGrady" YIELD edge as relationship |
```
Comment (yixinglu) [inline tests/tck/features/fetch/FetchEdges.strVid.feature:283]: Could we yield `properties(edge) as props` in fetch clause?

File: src/graph/planner/ngql/FetchEdgesPlanner.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                                  fetchCtx_->src,
+                                  fetchCtx_->type,
+                                  fetchCtx_->rank,
+                                  fetchCtx_->dst,
+                                  buildEdgeProps(),
+                                  {},
+                                  fetchCtx_->distinct);
+  getEdges->setInputVar(fetchCtx_->inputVarName);
+
+  subPlan.root = Filter::make(qctx, getEdges, emptyEdgeFilter());
```
Comment (yixinglu) [inline src/graph/planner/ngql/FetchEdgesPlanner.cpp:57]: Why need a filter here? Could do this same thing in GetEdges executor?

File: tests/tck/features/fetch/FetchEdges.strVid.feature (hunk 3 above)
Comment (nevermore3) [reply tests/tck/features/fetch/FetchEdges.strVid.feature:283]: ok

File: src/graph/planner/ngql/FetchEdgesPlanner.cpp (hunk 4 above)
Comment (nevermore3) [reply src/graph/planner/ngql/FetchEdgesPlanner.cpp:57]: filter out invalid edgeKeys or  Empty getEdges' result
for example:  FETCH PROP ON like. "Tim Duncan"-> "noExist" Yield edge as a
if there is no filter, will output a line with empty,  otherwise will output nothing. 
The current behavior is to output nothing

File: src/parser/TraverseSentences.h [hunk 5]
```diff
@@ -289,7 +289,7 @@ class FetchEdgesSentence final : public Sentence {
 
   YieldClause* yieldClause() const { return yieldClause_.get(); }
 
-  const std::string* edge() const { return edge_->front(); }
+  std::string edgeName() const { return *edge_->front(); }
```
Comment (Shylock-Hg) [inline src/parser/TraverseSentences.h:292]: return `const std::string&`

File: tests/tck/features/fetch/FetchEdges.intVid.feature [hunk 6]
```diff
@@ -176,45 +176,45 @@ Feature: Fetch Int Vid Edges
...
       | serve._src | serve._dst | serve._rank | serve.start_year | serve.end_year |
 
-  Scenario: Fetch prop Semantic Error
+  Scenario: Fetch prop Error
     When executing query:
       """
       FETCH PROP ON serve hash("Boris Diaw")->hash("Spurs") YIELD $^.serve.start_year
       """
-    Then a SemanticError should be raised at runtime:
+    Then a ExecutionError should be raised at runtime:
```
Comment (Shylock-Hg) [inline tests/tck/features/fetch/FetchEdges.intVid.feature:206]: Why change this?

File: tests/tck/features/fetch/FetchEdges.intVid.feature [hunk 7]
```diff
@@ -71,8 +71,8 @@ Feature: Fetch Int Vid Edges
   Scenario: Fetch prop works with pipeline
     When executing query:
       """
-      GO FROM hash('Boris Diaw') OVER serve YIELD serve._src AS src, serve._dst AS dst
-      | FETCH PROP ON serve $-.src->$-.dst YIELD serve.start_year, serve.end_year
+      GO FROM hash('Boris Diaw') OVER serve YIELD serve._src AS src, serve._dst AS dst |
```
Comment (Shylock-Hg) [inline tests/tck/features/fetch/FetchEdges.intVid.feature:74]: Don't modify the unrelated code.

File: tests/tck/features/fetch/FetchEdges.intVid.feature (hunk 6 above)
Comment (nevermore3) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:206]: validate $^.serve.start_year produce. ExecutorError

Comment (Shylock-Hg) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:206]: It's semantic error.

File: tests/tck/features/fetch/FetchEdges.intVid.feature (hunk 7 above)
Comment (nevermore3) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:74]: my formatting tool will report an error, if `|` at the forefront

Comment (Shylock-Hg) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:74]: Maybe you should update format tool.

File: tests/tck/features/fetch/FetchEdges.intVid.feature (hunk 6 above)
Comment (nevermore3) [reply tests/tck/features/fetch/FetchEdges.intVid.feature:206]: use duduceType produce ExecutorError

File: src/graph/validator/test/FetchEdgesTest.cpp [hunk 8]
```diff
@@ -359,6 +318,12 @@ TEST_F(FetchEdgesValidatorTest, FetchEdgesPropFailed) {
...
   ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD edge2.prop2"));
 
+  ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD edge"));
+  ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD vertex"));
+  ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD vertex as a, edge1.prop1"));
+  ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD vertex as b"));
+  ASSERT_FALSE(validate("FETCH PROP ON edge1 \"1\"->\"2\" YIELD path as a"));
+
   // notexist edge
   ASSERT_FALSE(validate("FETCH PROP ON not_exist_edge \"1\"->\"2\" YIELD not_exist_edge.prop1"));
```
Comment (Shylock-Hg) [inline src/graph/validator/test/FetchEdgesTest.cpp:328]: Add source property and dst property case.

Comment (nevermore3) [reply src/graph/validator/test/FetchEdgesTest.cpp:328]: done

File: src/parser/TraverseSentences.h (hunk 5 above)
Comment (nevermore3) [reply src/parser/TraverseSentences.h:292]: done

File: src/graph/planner/ngql/FetchEdgesPlanner.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#ifndef GRAPH_PLANNER_NGQL_FETCH_EDGES_H_
```
Comment (Shylock-Hg) [inline src/graph/planner/ngql/FetchEdgesPlanner.h:7]: pragma once

File: src/graph/validator/FetchEdgesValidator.h [hunk 10]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2020 vesoft inc. All rights reserved.
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
```
Comment (Shylock-Hg) [inline src/graph/validator/FetchEdgesValidator.h:1]: Don't modify this

Comment (nevermore3) [reply src/graph/validator/FetchEdgesValidator.h:1]: ignore it

File: src/graph/validator/test/FetchEdgesTest.cpp [hunk 11]
```diff
@@ -21,10 +21,10 @@ class FetchEdgesValidatorTest : public ValidatorTestBase {
 };
 
 TEST_F(FetchEdgesValidatorTest, FetchEdgesProp) {
-  auto src = VariablePropertyExpression::make(pool_.get(), "_VAR1_", kSrc);
+  auto src = ColumnExpression::make(pool_.get(), 0);
```
Comment (Shylock-Hg) [inline src/graph/validator/test/FetchEdgesTest.cpp:24]: Why change to ColumnExpression?

Comment (nevermore3) [reply src/graph/validator/test/FetchEdgesTest.cpp:24]: Because the column names of the first three columns in the dataset are fixed, we can use columnExpresion instead of VariableProperyExpression

File: src/graph/planner/ngql/FetchEdgesPlanner.h (hunk 9 above)
Comment (nevermore3) [reply src/graph/planner/ngql/FetchEdgesPlanner.h:7]: In my opinion, the effect is the same

Comment (Shylock-Hg) [reply src/graph/planner/ngql/FetchEdgesPlanner.h:7]: It's the new code style according to the doc.

**Review by Shylock-Hg — APPROVED** [review]
LGTM, but don't modify the unrelated behavior

**Review by CPWstatic — APPROVED** [review]
